corehq/motech/repeaters/management/commands/generate_repeater_report.py

```diff
@@ -0,0 +1,123 @@
+
+
+class Command(BaseCommand):
+    help = """
+    Pass Repeater ID along with domain and State(Optional) or
+    Pass a csv file path with a list of repeat records IDs to get a report(xlsx)
+    with final state and message for all attempts(if available)
+    """
+
+    def __init__(self, *args, **kwargs):
+        super(Command, self).__init__(*args, **kwargs)
+        # keep a counter of max attempts a repeat record has in the report
+        # to add those many headers later
+        self.max_attempts_in_sheet = 0
+        self.record_ids = []
+        self.wb = Workbook()
+        self.ws = self.wb.active
+        self.ws.append(['Record ID', 'Payload ID', 'State', 'Failure Reason'])
+
+    def add_arguments(self, parser):
+        parser.add_argument(
+            '--domain',
+            dest='domain',
+            help='Domain for which the repeater is passed'
+        )
+        parser.add_argument(
+            '--repeater_id',
+            dest='repeater_id',
+            help='ID for repeater for which the report is to be generated',
+        )
+        parser.add_argument(
+            '--state',
+            dest='state',
+            help='State of Record: CANCELLED, SUCCESS, PENDING, FAIL'
+        )
+        parser.add_argument(
+            '--records_file_path',
+            dest='records_file_path',
+            help='Path to a csv file with ids under header record_id'
+        )
+
+    def _add_row(self, repeat_record):
+        row = [repeat_record.get_id, repeat_record.payload_id, repeat_record.state, repeat_record.failure_reason]
+        repeat_record_attempts_count = len(repeat_record.attempts)
+        if repeat_record_attempts_count > self.max_attempts_in_sheet:
+            self.max_attempts_in_sheet = repeat_record_attempts_count
+        for attempt in repeat_record.attempts:
+            row.append(attempt.message)
+        self.ws.append(row)
+
+    def _add_header_for_attempts(self):
+        # find first empty header cell:
+        # iterate first row till highest column length which is highest column length over the sheet
+        # Even if there are no values in header, empty cells are inserted to match highest
+        # column length. So iterate over cells and find index for the first blank value
+        max_columns = self.ws.get_highest_column()
+        first_empty_index = max_columns
+        for i in range(1, max_columns + 1):
+            if not self.ws.cell(row=1, column=i).value:
+                first_empty_index = i
+                break
+        for i in range(0, self.max_attempts_in_sheet):
+            column_num = first_empty_index + i
+            self.ws.cell(row=1, column=column_num).value = "Attempt {index}".format(index=i + 1)
+
+    def _save_file(self, repeater_id, state):
+        if self.max_attempts_in_sheet > 0:
+            self._add_header_for_attempts()
+        file_name = "repeater_report_{repeater_id}_{state}_{timestamp}.xlsx".format(
```

Comment:
is there a reason you didn't just write this as a csv?

Reply:
I just find xl better to pass around to others, specially other stakeholders.  Plus, then there are no compatibility issues when they would open this file in excel and no surprises with fields having commas in them etc.